### ISSUE TITLE
search a specific job

### ISSUE THREAD

**Issue by felixmosh** [taskforcesh/bullmq#2135]
**Is your feature request related to a problem? Please describe.**
It can be create if queue can search a job with specific id, data etc...

**Describe the solution you'd like**
```js
   const jobs = await queue.search({jobData: 'contains some data'});
```

**Describe alternatives you've considered**
Seeking for the jobs on the redis it self

**Additional context**

**Comment by ccollie** [comment]
So I have (working) code in a personal project that does something similar. We can do something like:

```typescript
export interface FindJobsByFilterOptions {
  filter?: string;
  cursor?: string;
  status?: JobSearchStatus;
  pattern?: string;
  count?: number;
}

export interface FilteredJobsResult {
  cursor: string;
  total: number;
  current: number;
  jobs: Job[];
}

const filter = 'job.data.action == "create" && job.data.widget = "sprocket" && job.processTime < 100';

let { jobs, total, current, cursor } = await findJobsByFilter(queue, 
    { 
        filter, count: 10, cursor: null, pattern: 'job-*', status: "completed" 
    } 
)
if (jobs.length) {
    do {
        console.log(jobs);
        {jobs, cursor} = await findJobsByFilter(queue, { cursor } ); 
    } while(cursor)
}        

```
Currently it uses [jsep](https://github.com/EricSmekens/jsep) to parse `js` compatible expressions, determines which fields need to be fetched from `redis` and tries to as efficiently as possible retrieve the minimum data necessary to satisfy the predicate. 

Notice that it uses a cursor based approach to reduce the load on `redis` and provide a paging interface. Im not wedded to the expression syntax, and previous iterations used a `mongo` styled query document.

If Manuel thinks this fits within the philosophy of `bullmq`, I can clean it up and create a `PR`. Otherwise feel free to grab the code and plug away.

**Comment by kilobyte2007** [comment]
Are there any plans to develop this feature?

**Comment by felixmosh** [comment]
Once the underline lib (BullX) will add support for it.

For a meanwhile, you can up vote 
https://github.com/taskforcesh/bullmq/issues/2135

**Comment by kilobyte2007** [comment]
> Once the underline lib (BullX) will add support for it.
> 
> For a meanwhile, you can up vote #2135

Isn't this the underline lib?

**Comment by kilobyte2007** [comment]
> By "underline lib" I mean `bull` / `bullmq` which is used by the board to access all the data.

I get it, but this issue is open on the `bullmq` repo so I thought that was given :D

**Comment by leedia-tech** [comment]
Any news on this?

**Comment by bfbechlin** [comment]
This feature is really important. It would be amazing if it were prioritized 🙂

**Comment by ts-shivam** [comment]
A must have feature missing.

**Comment by pedropuppim** [comment]
This feature is much needed.

**Comment by faileon** [comment]
Any plans for this feature? Would be really nice to have :)

**Comment by ccollie** [comment]
In progress...

**Comment by edtimer** [comment]
Was thinking about this issue, what if we enable a toggle feature where a dedicated queue is created with searchable metadata that update on events. The toggle switch would enable the search ui so users can search. Since users send different data types we can use Record<string, any> . The limitations from this would be its not scalable  in my opinion with very large queues. Not sure this would be considered poking around indexes as @felixmosh mentioned

**Comment by manast** [comment]
The only way to make search practical with large queues would be to add some kind of index mechanism, so basically when adding jobs you would have a new option, something like:
```ts
myQueue.add( ..., { indexes: ["foo.bar", "bar.qux", ...] });
```
And every time you add a job the index gets updated for those fields in the job data, making searches on those fields very fast.

**Comment by Stosiu** [comment]
@manast @ccollie @roggervalf, I work on bull-board, where job search is still the most requested feature (felixmosh/bull-board#154, #361 and #476, with a bit over 100 reactions between them). Felix won't scan Redis from the dashboard, so bull-board is waiting on BullMQ for this, and I'd like to help get it moving.

I read through #3538. It works as a full scan, and ccollie mentioned it doesn't handle clustering yet, which is why it fails on Dragonfly (the job keys are built inside Lua instead of being passed in KEYS). Manast's comment here from April points the other way, towards an index maintained at write time. I looked at how other queues handle this and it lines up with that: Sidekiq and Faktory only pattern-scan their retry/dead sets, Horizon keeps a sorted set per tag, Temporal makes you register search attributes up front, and Oban and River rely on jsonb with GIN indexes.

Here is a shape I think could work, building on the `indexes` idea.

```ts
const queue = new Queue('orders', { search: { fields: ['name', 'data.tenantId'] } });
await queue.search({ field: 'data.tenantId', value: 'acme', limit: 50, cursor });
```

The fields would be declared on the queue and stored in the meta key, not passed on each `add`. Otherwise two producers that disagree leave the index silently incomplete, and a dashboard can read the meta key to know what's searchable. Values would be extracted in TS and passed to the add script, so there's no cjson on the hot path. Each entry is a ZADD into `{queue}:idx:<field>:<value>` scored by timestamp, which stays on one slot. State isn't indexed; it's resolved per page. On the Postgres backend the same declaration would map to a jsonb index.

### Keeping the index honest

Most of the work is here, so I'd rather agree on these before writing code.

**Which values a job was indexed under.** I'd store them on the job hash itself, as a small field such as `ix` holding `{"data.tenantId":"acme"}`. Removal can then ZREM exactly the right entries without decoding `data`, and it keeps working after the declaration changes, because the job records what it was indexed under rather than what the queue declares today.

**Removal paths.** Every place a job hash is deleted has to clean up its entries: `remove`, `removeOnComplete`/`removeOnFail` by count and by age, `clean`, `drain`, children removed with a parent, and `obliterate`, which can just drop `{queue}:idx:*`. If those all go through one Lua include this is a single hook; if they don't, it's the riskiest part of the change. Done this way, the index grows and shrinks with the jobs you actually keep, so there's no separate TTL or trimming to tune the way Horizon has.

**`updateData`.** With the `ix` field this is a diff: remove the old values, add the new ones, and rewrite `ix`, all in the same script. Without it you'd have to decode the old data in Lua.

**Adding a field to a queue that already has jobs.** Nothing already in the queue is indexed. I'd make backfill explicit, `queue.reindex({ field })`, walking the job ids in batches with a cursor, idempotent through ZADD NX, and skipping ids whose hash is already gone. It wouldn't run on a timer or on startup. The meta key records per field whether it's `complete` or `backfilling since <ts>`, so `search()` can tell the caller the result may be partial rather than silently missing older jobs. Jobs added while the backfill runs are indexed by the add path, so the two meet in the middle.

**Removing a field.** Drop it from meta, then delete its keys in batches. Keeping a small set per field listing its value keys (`{queue}:idx:<field>:values`) avoids having to SCAN for them.

**Producers that disagree or are older.** The meta key would be the single source of truth. A Queue constructed with a different `fields` list should throw rather than overwrite it, and changing the list goes through an explicit call. A producer on an older BullMQ that doesn't know about indexing will add unindexed jobs; the coverage flag above can't detect that, so it would have to be documented.

**Value normalisation.** Numbers and booleans are stringified, arrays index each element (so tags work for free), and missing or null paths get no entry. Very long values would either be capped or hashed. I'd keep case-folding opt-in per field.

**Prefix match.** This conflicts with one ZSET per value. A per-field lex ZSET of `value\0jobId` supports prefix search through ZRANGEBYLEX but loses time ordering. I'd start exact-only and add prefix per field later if people want it.

**Custom job ids and deduplication.** An `add` that hits an existing id or a dedupe key is a no-op today, and it must not write index entries either. Scheduler iterations are ordinary jobs and get indexed like any other.

**Filtering by state.** If state is resolved per page, a state-filtered search can return short pages. I'd rather document that than index state, which changes on every move and would double the write cost.

**Postgres.** Declaring a field there is DDL (an expression index or GIN). Whether BullMQ runs that itself, `CONCURRENTLY`, or leaves it to a migration is your call.

Free text over `data` could stay as a capped scan (newest N per state) based on the #3538 work, with the keys fixed for cluster. The index covers the common "find the jobs for this tenant/order/user" case, and the scan is only a fallback.

Would you take something along these lines? And ccollie, if you'd rather hand off #3538 or have me build on it instead of starting fresh, I'm happy either way. I can do the implementation once we agree on the shape.

**Comment by ccollie** [comment]
Hey @Stosiu, I apologize, but i've been super busy with family commitments. I'd be happy to collaborate on this. Let me know.
